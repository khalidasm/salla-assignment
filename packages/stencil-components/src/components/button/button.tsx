import { Component, h, Event, EventEmitter, Prop } from "@stencil/core";

@Component({
  tag: "button-component",
  styleUrl: "button.css",
  shadow: true,
})
export class Button {
  @Event() onClick: EventEmitter<void>;
  @Prop() variant?: "sm" | "normal" | "icon" = "normal";
  @Prop({mutable: true}) isDisabled?: boolean = false;
  @Prop({mutable: true}) isLoading?: boolean = false;
  @Prop({mutable: true}) showShimmer?: boolean = false;

  private handleClick = () => {
    if (this.onClick) {
      this.onClick.emit();
    }
  }

  render() {
    const getBaseClasses = () => {
      switch (this.variant) {
        case "icon":
          return "bg-transparent cursor-pointer";
        default:
          return "flex items-center justify-center w-full rounded-[6px] bg-[#004D5A] text-white hover:bg-[#004D5A]/80 hover:cursor-pointer transition-all duration-150 disabled:opacity-50 disabled:cursor-not-allowed";
      }
    };

    const getSizeClasses = () => {
      switch (this.variant) {
        case "sm":
          return "h-fit text-[10px] p-[7px]";
        case "icon":
          return "h-fit !p-0 w-fit";
        default:
          return "h-[45px] text-[14px] p-[10px]";
      }
    };

    if(this.showShimmer) {
      return <shimmer-component height={this.variant === "sm" ? "20px" : "45px"} width="100%"></shimmer-component>
    }

    return (
      <button onClick={this.handleClick} class={`${getBaseClasses()} ${getSizeClasses()}`} disabled={this.isDisabled || this.isLoading}>
        {this.isLoading ? <div class="h-5 w-5 animate-spin rounded-full border-2 border-white border-t-transparent" /> : <slot></slot>}
      </button>
    );
  }
}
