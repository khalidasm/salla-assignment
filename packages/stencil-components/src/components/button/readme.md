# button-component



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description | Type                         | Default    |
| ------------- | -------------- | ----------- | ---------------------------- | ---------- |
| `isDisabled`  | `is-disabled`  |             | `boolean`                    | `false`    |
| `isLoading`   | `is-loading`   |             | `boolean`                    | `false`    |
| `showShimmer` | `show-shimmer` |             | `boolean`                    | `false`    |
| `variant`     | `variant`      |             | `"icon" \| "normal" \| "sm"` | `"normal"` |


## Events

| Event     | Description | Type                |
| --------- | ----------- | ------------------- |
| `onClick` |             | `CustomEvent<void>` |


## Dependencies

### Used by

 - [cart-view-component](../cart-view)
 - [coupon-component](../coupon)
 - [coupon-input](../coupon-input)
 - [shipping-view-component](../shipping-view)

### Depends on

- [shimmer-component](../shimmer)

### Graph
```mermaid
graph TD;
  button-component --> shimmer-component
  cart-view-component --> button-component
  coupon-component --> button-component
  coupon-input --> button-component
  shipping-view-component --> button-component
  style button-component fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
