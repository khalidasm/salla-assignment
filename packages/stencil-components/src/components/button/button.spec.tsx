import { newSpecPage } from '@stencil/core/testing';
import { Button } from './button';

describe('button-component', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<button-component>Click me</button-component>',
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with small variant', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<button-component variant="sm">Small Button</button-component>',
    });
    expect(page.root).toMatchSnapshot();
    const button = page.root.shadowRoot.querySelector('button');
    expect(button.className).toContain('h-fit');
    expect(button.className).toContain('text-[10px]');
    expect(button.className).toContain('p-[7px]');
  });

  it('renders with icon variant', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<button-component variant="icon">Icon</button-component>',
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders in disabled state', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<button-component is-disabled>Disabled</button-component>',
    });
    expect(page.root).toMatchSnapshot();
    const button = page.root.shadowRoot.querySelector('button');
    expect(button.className).toContain('disabled');
  });

  it('renders in loading state', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<button-component is-loading>Loading</button-component>',
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with shimmer', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<button-component show-shimmer>Shimmer</button-component>',
    });
    expect(page.root).toMatchSnapshot();
  });

  it('emits click events', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: '<button-component>Click me</button-component>',
    });

    const onClickSpy = jest.fn();
    page.root.addEventListener('onClick', onClickSpy);

    const button = page.root.shadowRoot.querySelector('button');
    button.click();

    expect(onClickSpy).toHaveBeenCalled();
  });
}); 