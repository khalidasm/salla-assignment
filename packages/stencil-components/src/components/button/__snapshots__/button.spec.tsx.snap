// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button-component renders in disabled state 1`] = `
<button-component is-disabled="">
  <template shadowrootmode="open">
    <button class="bg-[#004D5A] disabled:cursor-not-allowed disabled:opacity-50 duration-150 flex h-[45px] hover:bg-[#004D5A]/80 hover:cursor-pointer items-center justify-center p-[10px] rounded-[6px] text-[14px] text-white transition-all w-full" disabled="">
      <slot></slot>
    </button>
  </template>
  Disabled
</button-component>
`;

exports[`button-component renders in loading state 1`] = `
<button-component is-loading="">
  <template shadowrootmode="open">
    <button class="bg-[#004D5A] disabled:cursor-not-allowed disabled:opacity-50 duration-150 flex h-[45px] hover:bg-[#004D5A]/80 hover:cursor-pointer items-center justify-center p-[10px] rounded-[6px] text-[14px] text-white transition-all w-full" disabled="">
      <div class="animate-spin border-2 border-t-transparent border-white h-5 rounded-full w-5"></div>
    </button>
  </template>
  Loading
</button-component>
`;

exports[`button-component renders with default props 1`] = `
<button-component>
  <template shadowrootmode="open">
    <button class="bg-[#004D5A] disabled:cursor-not-allowed disabled:opacity-50 duration-150 flex h-[45px] hover:bg-[#004D5A]/80 hover:cursor-pointer items-center justify-center p-[10px] rounded-[6px] text-[14px] text-white transition-all w-full">
      <slot></slot>
    </button>
  </template>
  Click me
</button-component>
`;

exports[`button-component renders with icon variant 1`] = `
<button-component variant="icon">
  <template shadowrootmode="open">
    <button class="!p-0 bg-transparent cursor-pointer h-fit w-fit">
      <slot></slot>
    </button>
  </template>
  Icon
</button-component>
`;

exports[`button-component renders with shimmer 1`] = `
<button-component show-shimmer="">
  <template shadowrootmode="open">
    <shimmer-component height="45px" width="100%"></shimmer-component>
  </template>
  Shimmer
</button-component>
`;

exports[`button-component renders with small variant 1`] = `
<button-component variant="sm">
  <template shadowrootmode="open">
    <button class="bg-[#004D5A] disabled:cursor-not-allowed disabled:opacity-50 duration-150 flex h-fit hover:bg-[#004D5A]/80 hover:cursor-pointer items-center justify-center p-[7px] rounded-[6px] text-[10px] text-white transition-all w-full">
      <slot></slot>
    </button>
  </template>
  Small Button
</button-component>
`;
